File: SeLeAs/Python/LeetCode 75/findCircleNum.py
from typing import List

class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        def dfs(i):
            visited.add(i)
            for j in range(n):
                if isConnected[i][j] and j not in visited:
                    dfs(j)
        
        n = len(isConnected)
        visited = set()
        province = 0
        
        for i in range(n):
            if i not in visited:
                dfs(i)
                province += 1
                
        return province

File: LeetCode/Study_Plan/LeetCode_75/findCircleNum.java
import java.util.HashSet;
import java.util.Set;

public class findCircleNum{
class Solution {
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        Set<Integer> visited = new HashSet<>();
        int province = 0;

        for (int i = 0; i < n; i++) {
            if (!visited.contains(i)) {
                dfs(i, isConnected, visited, n);
                province++;
            }
        }

        return province;
    }

    private void dfs(int i, int[][] isConnected, Set<Integer> visited, int n) {
        visited.add(i);
        for (int j = 0; j < n; j++) {
            if (isConnected[i][j] == 1 && !visited.contains(j)) {
                dfs(j, isConnected, visited, n);
            }
        }
    }
}
}